docs/source/user-guide/concepts/expressions-and-contexts.md

```diff
@@ -0,0 +1,192 @@
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:df"
+```
+
+### `select`
+
+The selection context `select` applies expressions over columns.
+The context `select` may produce new columns that are aggregations, combinations of other columns, or literals:
+
+{{code_block('user-guide/concepts/expressions','select-1',['select'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:select-1"
+```
+
+The expressions in a context `select` must produce series that are all the same length or have length 1.
+Series that have length 1 will be broadcast to match the length of the remaining series.
+Literals, like the number used above, are treated as series of length 1 and are also broadcast.
+
+Note that broadcasting can also occur within expressions.
+For instance, consider the expression below:
+
+{{code_block('user-guide/concepts/expressions','select-2',['select'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:select-2"
+```
+
+Both the subtraction and the division use broadcasting within the expression because the subexpressions that compute the mean and the standard deviation evaluate to single values.
+
+The context `select` is very flexible and powerful and allows you to evaluate arbitrary expressions independent of, and in parallel to, each other.
+This is also true of the other contexts that we will see next.
+
+### `with_columns`
+
+The context `with_columns` is very similar to the context `select`.
+The main difference between the two is that the context `with_columns` retains the original columns and adds new ones according to its input expressions, whereas the context `select` drops the original columns:
+
+{{code_block('user-guide/concepts/expressions','with_columns-1',['with_columns'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:with_columns-1"
+```
+
+### `filter`
+
+The context `filter` filters the rows of a dataframe based on one or more expressions that evaluate to the Boolean data type.
+
+{{code_block('user-guide/concepts/expressions','filter-1',['filter'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:filter-1"
+```
+
+### `group_by` and aggregations
+
+In the context `group_by`, rows are grouped according to the unique values of the grouping expressions.
+You can then apply expressions to the resulting groups, which may be of variable lengths.
+
+When using the context `group_by`, you can use an expression to compute the groupings dynamically:
+
+{{code_block('user-guide/concepts/expressions','group_by-1',['group_by'])}}
```

Comment:
Can we improve the autoformat here. I think we need a trailing comma somewhere so that it will format to multiple lines.